### ISSUE TITLE
"cat /dev/random | base64 | head -n10" crashes Elvish

### ISSUE THREAD

**Issue by xiaq** [elves/elvish#994]
This line does not handle `e.Cause == nil`:

https://github.com/elves/elvish/blob/52bd0aef41819e151d359370c7a0e435d12c9925/pkg/eval/exception.go#L60

**Comment by xiaq** [comment]
Fixing this per se is straightforward but I want to spend some time constructing a minimal example and there seems to be bigger problems behind this.

**Comment by zzamboni** [comment]
This might be OS-specific, on macOS I cannot reproduce it:
```
[~]─> cat /dev/random | base64 | head -n10                                                                                                                                                
E+ObwttEOz3iFLSutNpAUIfoCPqLPpQbcKQIoIHu8bs5N/pBCC3mYWOdYmLS8G1ioG8U//EOU1lJ
bD8+mnfsDP8t+9clT9kQRsdge23Q3I3ojsaPe5SaYKDIHmpfY6xG053vsPoLvdOciPjczE/nfZdo
j5hyAB2z361S06hf2frMFBdnF1hOflrVc+JfIWP4BjYIesZVAbF1qu8JFHfBzgI/opiHcvNqZXEN
VMXrYsqNk9US9bhoQCm77CBkSrFUzIKPalx769/buATed+HzahaVkh3nFeaIju/Izg8EVtTd7Fi+
HBxcOcTthN8lYXgf/cucYJwF1DmUVzHu+Y2scEYzf0rPAg7oBEZTbLELzq66wkAKjc28/aRbXhu6
hywgssTZFw/QxF0mPLdsHDMRwDw1Axeb3bN7bEiWeJMbQkKWW8eXmpxKNJNw1QWBR8f7sN7CYUWP
LB6N6/kgEESX7DmyQB5kxEIkOMxXAcDaYzfnuORTZah8j2O0cvyw/uVvvNANQ00sHORqiHvOP0MI
jBZfAdYhFzeFXVIZIKyVUSLxy2VvJBqIhcNKrbfn7fDQ1dbHcwU/5ruFWySQ/hGLnGLLDFHgtGV7
OtK17szNd3tJAEakBI2CX70Zcj3CWWRzCKY5N7SDFSAnP2fKJO1cXCKzp//ALNsmuu/rif0HX5Pp
dD9nZ5LoQ+mjLYejRp0qKN84o9FFl2JhuvaIHmSZaQp0hZ9FhpqiMWAxJPZprbnbiwyTdxskylKp
Exception: base64 killed by signal broken pipe
[tty 20], line 1: cat /dev/random | base64 | head -n10
```

**Comment by hanche** [comment]
@zzamboni What is your `base64` program? For me, the command as shown does not terminate until I interrupt it, because `base64` by default does not insert linebreaks. If I run the pipeline with `base64 -b 60`, I do get the crash. This is on the latest Catalina (10.15.4).

```
⬥ search-external base64
⮕ /usr/bin/base64
```

I get many strange variants on this, though. It does not look like `base64` is essential:

```
⬥ cat /dev/random | head -c 1024 > tttt
Exception: cat killed by signal broken pipe
[tty 6], line 1: cat /dev/random | head -c 1024 > tttt
⬥ cat /dev/random | cat | head -c 1024 > tttt
# crashes
```

**Comment by zzamboni** [comment]
@hanche good point - I realized I am using by default `base64` from the GNU coreutils package:
```
[~]─> search-external base64
▶ /usr/local/opt/coreutils/libexec/gnubin/base64
```
If I explicitly insert `/usr/bin/base64` in the command, I still don't get a crash, but the Terminal hangs up and I have to force-quit it.

**Comment by krader1961** [comment]
TLDR; You can reproduce the panic on macOS and Linux by doing this:

```
sh -c 'exit 1' | sh -c 'exit 2' | cat
```

Note that `base64 </dev/random | head -n10` results in elvish correctly reporting a sigpipe exception for the `base64` command. Also, there is no need for `head -n10`. Replacing that with the following trivial POSIX script results in the same panic when doing `cat /dev/random | base64 | ~/trivial`:

```
#!/bin/sh
exec <&-
sleep 1
```

In fact, using that trivial replacement for the `head -n10` causes the same panic on macOS without the need to do `base64 -b76`. In fact, the `base64` isn't needed on either macOS or Linux. It's sufficient to do `cat /dev/random | cat | ~/trivial`.

Using `strace` shows the elvish shell received a `SIGCHLD` immediately prior to the panic for a pid not involved in the pipeline. That process was the elvish daemon started before I traced the shell and the exit status of that process was zero. While the panicing shell received a `SIGCHLD` for the daemon it had not executed a `wait4()` at the time of the panic. However, I suspect the panic was already in progress at that point and elvish was simply terminating the daemon before emitting the panic stacks.

Shortly before that it received a `SIGCHLD` for the `cat /dev/random` process and had completed a `wait4()` call on the process which showed the cat died due to `SIGPIPE`. Same for the `cat` in the middle of the pipeline.

This is a simpler way to cause the panic which confirms my suspicion this has nothing to do with SIGPIPE:

```
sh -c 'exit 1' | sh -c 'exit 2' | cat
```

It's pretty clear at this point the problem is caused by reporting an exception for more than one process in a pipeline.

**Comment by krader1961** [comment]
@xiaq wrote:

> I want to spend some time constructing a minimal example and there seems to be bigger problems behind this.

The minimal example, that does not involve issues surrounding SIGPIPE, is this:

```
sh -c 'exit 1' | sh -c 'exit 2' | cat
```

It seems that exceptions raised by more than one process in a pipeline isn't handled correctly. Is there perhaps a global var involved in throwing an elvish exception for an external command?
